### PR TITLE
EU-89

### PR DESCRIPTION
**Pull request by victor-nuzhniy**
Waiting for review.

### CODE REVIEW

File: components/exception_handlers.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from components.enums import JSENDStatus
+from components.exceptions import BackendException
+from settings import Settings
+from starlette.exceptions import HTTPException as StarletteHTTPException
+
+
+def backend_exception_handler(request: Request, exc: BackendException) -> JSONResponse:
+    """Return result from Back-end exception."""
+    response = JSONResponse(content=exc.dict())
+    response.status_code = exc.code
```
Comment (Kostiantyn-Salnykov) [inline components/exception_handlers.py:15]: You can use directly `status_code` inside initialization of JSONResponse.

File: components/exception_handlers.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        )
+    response = JSONResponse(
+        content={
+            "status": JSENDStatus.FAIL,
+            "data": modified_details,
+            "message": "Validation error.",
+            "code": status.HTTP_422_UNPROCESSABLE_ENTITY,
+        }
+    )
+    response.status_code = status.HTTP_422_UNPROCESSABLE_ENTITY
```
Comment (Kostiantyn-Salnykov) [inline components/exception_handlers.py:54]: You can use directly `status_code` inside initialization of JSONResponse.

File: handlers/authorization/registration.py [hunk 3]
```diff
@@ -12,55 +12,66 @@
...
+async def registation(user: RegistrationIn):  # TODO syntax error
+
     RegistrationIn(
         token=user.token,
         email=user.email,
         password=user.password,
         password_re_check=user.password_re_check)
 
-    query = select(OneTimeToken).where(OneTimeToken.token == user.token)
+    query = select(OneTimeToken).where(OneTimeToken.c.token == user.token)
```
Comment (Kostiantyn-Salnykov) [inline handlers/authorization/registration.py:32]: Re-check this `.c` call, I gueess @YehorVoitenko removed it while refactor Models to declarative base.

File: schemas/student.py [hunk 4]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
-    def validate_full_name(value):
-        full_name = value.split()
-        if not full_name:
-            raise ValueError("Прізвище та ім'я студента обов'язкові до заповнення!")
-        elif len(full_name) < 2:
-            raise ValueError("Прізвище та ім'я студента обов'язкові до заповнення!")
+    def validate_full_name(value):  # TODO may be there is need to use staticmethod decorator
+        full_name = value.split()   # TODO use hint or so
+        if not full_name or len(full_name) < 2:
+            raise BackendException(message="The student's name and surname are mandatory!")
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:30]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 5]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         return value
     
     @validator('telephone_number')
-    def validate_telephone_number(value):
+    def validate_telephone_number(value):   # TODO may be there is need to use staticmethod decorator
         if not value:
-            raise ValueError('Телефонний номер не може бути порожнім!')
+            raise BackendException(message='The phone number cannot be empty!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:36]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 6]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         elif len(str(value)) != 12:
-            raise ValueError('Телефонний номер має містити в собі 12 цифр!')
+            raise BackendException(message='The phone number must contain 12 digits!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:38]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 7]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         return value
 
     @validator('course_id')
-    def validate_course_id(value):
+    def validate_course_id(value):  # TODO may be there is need to use staticmethod decorator
         if not value:
-            raise ValueError('Курс не може бути порожнім!')
+            raise BackendException(message='The course cannot be empty!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:44]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 8]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         elif value not in range(1, 7):
-            raise ValueError('Курс моє бути між 1 та 6!')
+            raise BackendException(message='The course must be between 1 and 6!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:46]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 9]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         return value
 
     @validator('speciality_id')
     def validate_speciality_id(value):
         if not value:
-            raise ValueError('Cпеціальність не може бути порожньою!')
+            raise BackendException(message='The specialty cannot be empty!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:52]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 10]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         return value
 
     @validator('faculty_id')
     def validate_faculty_id(value):
         if not value:
-            raise ValueError('Факультет не може бути порожнім')
+            raise BackendException(message='The faculty cannot be empty!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:58]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 11]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
         return value
 
     @validator('gender')
     def validate_gender(value):
         exists_genders = ['Ч', 'М']
         if not value: 
-            raise ValueError('Стать студента не може бути порожня')
-        if value.upper() not in exists_genders:
-            raise ValueError('Оберіть стать із запропонованого списку')
+            raise BackendException(message='The student gender cannot be empty!')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:65]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/student.py [hunk 12]
```diff
@@ -23,49 +24,47 @@ class CreateStudentIn(BaseModel):
...
+        if value.upper() not in exists_genders:     # TODO use hint or so
+            raise BackendException(message='Choose your gender from the list provided.')
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:67]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 13]
```diff
@@ -45,26 +45,23 @@ def validate_email(cls, v):
...
+            message = "The email address cannot be empty."
 
         elif not re.fullmatch(regex, v):
-            message = f"Невірний формат адресу електронної пошти: {v}."
+            message = f"Invalid email address format: {v}."
 
         if message:
-            raise ValueError(message)
-
+            raise BackendException(message=message)
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:54]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 14]
```diff
@@ -45,26 +45,23 @@ def validate_email(cls, v):
...
 
     @validator('password_re_check')
-    def validate_password(cls, v, values):
+    def validate_password(cls, v, values):  # TODO may be there is need to use classmethod decorator
         password = values.get('password')
 
         if not password or not v:
-            raise ValueError('Паролі не можуть бути порожніми')
-
+            raise BackendException(message='Passwords cannot be empty.')
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:62]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 15]
```diff
@@ -45,26 +45,23 @@ def validate_email(cls, v):
...
         if password != v:
-            raise ValueError('Введені паролі не співпадають')
-
+            raise BackendException(message='The entered passwords do not match.')
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:64]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 16]
```diff
@@ -96,24 +93,21 @@ def validate_email(cls, v):
...
         if not v:
-            message = "Електронний адрес не може бути порожнім"
+            message = "The email address cannot be empty."
         elif not re.fullmatch(regex, v):
-            message = f"Невірний формат адреси електронної пошти: {v}."
+            message = f"Invalid email address format: {v}."
         if message:
-            raise ValueError(message)
-        
+            raise BackendException(message=message)
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:100]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 17]
```diff
@@ -96,24 +93,21 @@ def validate_email(cls, v):
...
         return v
 
     @validator('password_re_check')
     def validate_password(cls, v, values):
         password = values.get('password')
 
         if not password or not v:
-            raise ValueError('Паролі не можуть бути порожніми')
-
+            raise BackendException(message='Passwords cannot be empty.')
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:108]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user.py [hunk 18]
```diff
@@ -96,24 +93,21 @@ def validate_email(cls, v):
...
         if password != v:
-            raise ValueError('Введені паролі не співпадають')
-
+            raise BackendException(message='The entered passwords do not match.')
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:110]: Don't use `BackendException` inside Pydantic schemas validation. it should be handled by another handler: `validation_exception_handler`.

File: schemas/user_request.py [hunk 19]
```diff
@@ -74,12 +74,12 @@ class UserRequestReviewIn(BaseModel):
     bed_place_id: int = None
 
     @validator('status_id')
-    def validate_status_id(cls, v):
-        if v not in [1,2]:
-            message = "Заяву можливо тільки ухвалити або відхилити."
-            raise ValueError(message)
+    def validate_status_id(cls, v):  # TODO may be there is need to use classmethod decorator
+        if v not in [1, 2]:
+            raise BackendException(message="The application can only be approved or rejected.")
```
Comment (Kostiantyn-Salnykov) [inline schemas/user_request.py:79]: Don't use BackendException inside Pydantic schemas validation. it should be handled by another handler: validation_exception_handler.

File: schemas/user_request.py [hunk 20]
```diff
@@ -50,10 +51,9 @@ class CancelRequestIn(BaseModel):
     status_id: int
 
     @validator('status_id')
-    def validate_status_id(cls, v):
+    def validate_status_id(cls, v):  # TODO may be there is need to use classmethod decorator
         if v != 4:
-            message = "Заяву можливо тільки скасувати."
-            raise ValueError(message)
+            raise BackendException(message="The application can only be canceled.")
```
Comment (Kostiantyn-Salnykov) [inline schemas/user_request.py:56]: Don't use BackendException inside Pydantic schemas validation. it should be handled by another handler: validation_exception_handler.

File: schemas/user.py [hunk 21]
```diff
@@ -1,15 +1,15 @@
...
 from pydantic import BaseModel, validator
+from components.exceptions import BackendException
 
 
 class UsersListViewOut(BaseModel):
 
     user_id: int
     login: str
-    last_vist: datetime = None
+    last_vist: datetime = None      # TODO syntax error
```
Comment (Kostiantyn-Salnykov) [inline schemas/user.py:12]: Change it to `last_visit` if this field has no syntax errors inside declaration of view (check main.sql, and ...views.py for that)

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

**Review by anton-serozhechkin — CHANGES_REQUESTED** [review]

**Review by anton-serozhechkin — DISMISSED** [review]

File: handlers/authorization/registration.py (hunk 3 above)
Comment (victor-nuzhniy) [reply handlers/authorization/registration.py]: ok

File: components/exception_handlers.py (hunk 1 above)
Comment (victor-nuzhniy) [reply components/exception_handlers.py]: ok

File: components/ecxeption_handlers.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
+from fastapi import Request, status
```
[the remark this answers is not in the document]
Comment (victor-nuzhniy) [reply components/ecxeption_handlers.py]: ok

File: schemas/user.py (hunk 21 above)
Comment (victor-nuzhniy) [reply schemas/user.py]: ok

**Review by anton-serozhechkin — DISMISSED** [review]

File: components/exceptions.py [hunk 23]
```diff
@@ -19,7 +19,7 @@ def __init__(
         status: JSENDStatus = JSENDStatus.FAIL,
         data: typing.Union[None, int, str, list, dict] = None,
         message: str,
-        code: int = http_status.HTTP_400_BAD_REQUEST
+        code: int = http_status.HTTP_422_UNPROCESSABLE_ENTITY
```
Comment (Kostiantyn-Salnykov) [inline components/exceptions.py:22]: Should be 400 status code by default.

File: schemas/student.py [hunk 24]
```diff
@@ -23,49 +22,47 @@ class CreateStudentIn(BaseModel):
...
 
     @validator('full_name')
-    def validate_full_name(value):
-        full_name = value.split()
-        if not full_name:
-            raise ValueError("Прізвище та ім'я студента обов'язкові до заповнення!")
-        elif len(full_name) < 2:
-            raise ValueError("Прізвище та ім'я студента обов'язкові до заповнення!")
+    def validate_full_name(value):  # TODO may be there is need to use staticmethod decorator
+        full_name = value.split()   # TODO use hint or so
```
Comment (Kostiantyn-Salnykov) [inline schemas/student.py:26]: Sure, you can use type hints for (something like that):
```python
def validate_full_name(value: str):
    full_name: List[str, str] = value.split()
```

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

**Review by Kostiantyn-Salnykov — APPROVED** [review]

File: schemas/student.py (hunk 24 above)
Comment (anton-serozhechkin) [reply schemas/student.py]: not resolved but ok for now

**Review by anton-serozhechkin — APPROVED** [review]

**Pull request merged**